nexus/src/db/queries/external_ip.rs

```diff
@@ -0,0 +1,627 @@
+                ),
+            }
+        );
+        context.success().await;
+    }
+
+    #[tokio::test]
+    async fn test_next_external_ip_out_of_order_allocation_ok() {
+        let context = TestContext::new(
+            "test_next_external_ip_out_of_order_allocation_ok",
+        )
+        .await;
+        // Need a larger range, since we're currently limited to the whole port
+        // range for each external IP.
+        let range = IpRange::try_from((
+            Ipv4Addr::new(10, 0, 0, 1),
+            Ipv4Addr::new(10, 0, 0, 3),
+        ))
+        .unwrap();
+        context.create_ip_pool("p0", range).await;
+
+        // TODO-completess: Implementing Iterator for IpRange would be nice.
+        let addresses = [
+            Ipv4Addr::new(10, 0, 0, 1),
+            Ipv4Addr::new(10, 0, 0, 2),
+            Ipv4Addr::new(10, 0, 0, 3),
+        ];
+        let ports = (0..super::MAX_PORT).step_by(super::NUM_SOURCE_NAT_PORTS);
+        let mut external_ips = itertools::iproduct!(addresses, ports);
+
+        // Allocate two addresses
+        let mut ips = Vec::with_capacity(2);
+        for (expected_ip, expected_first_port) in external_ips.clone().take(2) {
+            let instance_id = Uuid::new_v4();
+            let ip = context
+                .db_datastore
+                .allocate_instance_external_ip(&context.opctx, instance_id)
+                .await
+                .expect("Failed to allocate instance external IP address");
+            assert_eq!(ip.ip.ip(), expected_ip);
+            assert_eq!(ip.first_port.0, expected_first_port as u16);
+            let expected_last_port = (expected_first_port
+                + (super::NUM_SOURCE_NAT_PORTS - 1) as i32)
+                as u16;
+            assert_eq!(ip.last_port.0, expected_last_port);
+            ips.push(ip);
+        }
+
+        // Release the first
+        context
+            .db_datastore
+            .deallocate_instance_external_ip(&context.opctx, ips[0].id)
+            .await
+            .expect("Failed to release the first external IP address");
+
+        // Allocate a new one, ensure it's the same as the first one we
+        // released.
+        let instance_id = Uuid::new_v4();
+        let ip = context
+            .db_datastore
+            .allocate_instance_external_ip(&context.opctx, instance_id)
+            .await
+            .expect("Failed to allocate instance external IP address");
+        assert_eq!(
+            ip.ip, ips[0].ip,
+            "Expected to reallocate external IPs sequentially"
+        );
+        assert_eq!(
+            ip.first_port, ips[0].first_port,
+            "Expected to reallocate external IPs sequentially"
+        );
+        assert_eq!(
+            ip.last_port, ips[0].last_port,
+            "Expected to reallocate external IPs sequentially"
+        );
+
+        // Allocate one more, ensure it's the next chunk after the second one
+        // from the original loop. At this point
```

Comment:
At this point... ?

Reply:
Hehe really lost my train of thought there. Thanks, will fix.